lib/ae_mdw/sync/server.ex

```diff
@@ -1,181 +1,353 @@
-      height
-      |> :aec_db.find_headers_at_height()
-      |> Enum.filter(&(:aec_headers.type(&1) == :key))
+    actions = [{:next_event, :internal, :check_sync}]
+    new_gens_per_min = calculate_gens_per_min(prev_gens_per_min, gens_per_min)
 
-    new_state =
-      if length(header_candidates) > 1 do
-        # FORK!
-        main_header =
-          Enum.find_value(header_candidates, fn header ->
-            {:ok, header_hash} = :aec_headers.hash_header(header)
+    new_state_data = %__MODULE__{
+      state_data
+      | db_state: new_mem_state,
+        gens_per_min: new_gens_per_min
+    }
 
-            # COSTLY!
-            :aec_chain_state.hash_is_in_main_chain(header_hash) && header
-          end)
+    {:next_state, :idle, new_state_data, actions}
+  end
 
-        fork_height = :aec_headers.height(main_header)
+  def handle_event({:call, from}, :gens_per_min, _state, %__MODULE__{gens_per_min: gens_per_min}) do
+    actions = [{:reply, from, gens_per_min}]
 
-        %__MODULE__{
-          s
-          | event_handler: EventHandler.process_event!({:fork, fork_height}, event_handler)
-        }
-      else
-        s
-      end
+    {:keep_state_and_data, actions}
+  end
+
+  def handle_event({:call, from}, :syncing?, state, _data) do
+    syncing? = state != :initialized and state != :stopped
+    actions = [{:reply, from, syncing?}]
 
-    {:noreply, new_state}
+    {:keep_state_and_data, actions}
   end
 
-  def handle_info({:gproc_ps_event, :top_changed, %{info: %{block_type: :micro}}}, s),
-    do: {:noreply, s}
+  def handle_event(:info, :restart_sync, :stopped, state_data) do
+    actions = [{:next_event, :internal, :check_sync}]
+
+    {:next_state, :idle, %__MODULE__{state_data | restarts: 0}, actions}
+  end
 
-  def handle_info(
+  def handle_event(
+        :info,
         {:DOWN, _ref, :process, pid, reason},
-        %__MODULE__{event_handler: event_handler} = state
-      ) do
-    case EventHandler.process_event({:pid_down, pid, reason}, event_handler) do
-      {:ok, next_event_handler} ->
-        {:noreply, %__MODULE__{state | event_handler: next_event_handler}}
+        {:syncing_db, pid},
+        %__MODULE__{restarts: restarts} = state_data
+      )
+      when restarts < @max_restarts do
+    Log.info("DB Sync.Server error: #{inspect(reason)}")
 
-      :stop ->
-        :aec_events.unsubscribe(:chain)
-        :aec_events.unsubscribe(:top_changed)
+    actions = [{:next_event, :internal, :check_sync}]
 
-        :timer.apply_after(@retry_time, __MODULE__, :start_sync, [])
+    {:next_state, :idle, %__MODULE__{state_data | restarts: restarts + 1}, actions}
+  end
 
-        {:noreply, %__MODULE__{state | syncing?: false}}
-    end
+  def handle_event(
+        :info,
+        {:DOWN, _ref, :process, pid, reason},
+        {:syncing_mem, pid},
+        %__MODULE__{restarts: restarts} = state_data
+      )
+      when restarts < @max_restarts do
+    Log.info("Mem Sync.Server error: #{inspect(reason)}")
```

Comment:
After monitoring it's better use symmetrical calls including the `demonitor`.

Reply:
Done, implemented using Task.Supervisor as suggested